File: app/metrics.py
"""Aggregate statistics over a list of RequestMetrics: percentiles for
latency and throughput. Pure functions, no I/O -- easy to unit-test and to
call from both the API layer and any future CLI report.
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Optional

from app.models import RequestMetrics


def percentile(values: list[float], p: float) -> Optional[float]:
    """Nearest-rank percentile. p is 0-100. Returns None on empty input."""
    if not values:
        return None
    s = sorted(values)
    if len(s) == 1:
        return s[0]
    rank = math.ceil(p / 100 * len(s)) - 1
    rank = min(max(rank, 0), len(s) - 1)
    return s[rank]


@dataclass
class Summary:
    n: int
    n_ok: int
    n_error: int

    ttft_p50: Optional[float]
    ttft_p95: Optional[float]
    ttft_p99: Optional[float]

    total_p50: Optional[float]
    total_p95: Optional[float]
    total_p99: Optional[float]

    decode_tps_mean: Optional[float]

    def as_dict(self) -> dict:
        return {
            "n": self.n,
            "n_ok": self.n_ok,
            "n_error": self.n_error,
            "ttft_p50": self.ttft_p50,
            "ttft_p95": self.ttft_p95,
            "ttft_p99": self.ttft_p99,
            "total_p50": self.total_p50,
            "total_p95": self.total_p95,
            "total_p99": self.total_p99,
            "decode_tps_mean": self.decode_tps_mean,
        }


def summarize(metrics: list[RequestMetrics]) -> Summary:
    """Summarize a batch of results, excluding warmup requests. Errors are
    counted but excluded from the latency/throughput percentiles so a
    handful of failures don't corrupt the timing picture.
    """
    scored = [m for m in metrics if not m.is_warmup]
    ok = [m for m in scored if m.ok]
    errors = [m for m in scored if not m.ok]

    ttfts = [m.ttft_s for m in ok if m.ttft_s is not None]
    totals = [m.total_s for m in ok if m.total_s is not None]
    tps = [m.decode_tokens_per_s for m in ok if m.decode_tokens_per_s is not None]

    return Summary(
        n=len(scored),
        n_ok=len(ok),
        n_error=len(errors),
        ttft_p50=percentile(ttfts, 50),
        ttft_p95=percentile(ttfts, 95),
        ttft_p99=percentile(ttfts, 99),
        total_p50=percentile(totals, 50),
        total_p95=percentile(totals, 95),
        total_p99=percentile(totals, 99),
        decode_tps_mean=(sum(tps) / len(tps)) if tps else None,
    )


def compute_decode_rate(completion_tokens: Optional[int], ttft_s: Optional[float], total_s: Optional[float]) -> Optional[float]:
    """Tokens/sec during the decode phase only (excludes TTFT), so a slow
    prompt-eval doesn't drag down the apparent generation speed.
    """
    if not completion_tokens or total_s is None or ttft_s is None:
        return None
    decode_s = total_s - ttft_s
    if decode_s <= 0:
        return None
    return completion_tokens / decode_s

File: lib/metrics.ts
// Aggregate statistics over a batch of RequestMetrics: percentiles for
// latency, mean decode throughput. Pure functions -- no I/O.

import type { RequestMetrics, RunSummary } from "@/lib/types";
import { parseChecks } from "@/lib/types";

export function percentile(values: number[], p: number): number | null {
  if (values.length === 0) return null;
  const sorted = [...values].sort((a, b) => a - b);
  if (sorted.length === 1) return sorted[0]!;
  const rank = Math.min(Math.max(Math.ceil((p / 100) * sorted.length) - 1, 0), sorted.length - 1);
  return sorted[rank]!;
}

/** Summarizes one (endpoint, case) group, excluding warmup requests.
 * Errors are counted but excluded from latency/throughput percentiles so
 * a handful of failures don't corrupt the timing picture.
 */
export function summarize(endpointId: number, caseId: number, metrics: RequestMetrics[]): RunSummary {
  const scored = metrics.filter((m) => !m.is_warmup);
  const ok = scored.filter((m) => m.ok);
  const errors = scored.filter((m) => !m.ok);

  const ttfts = ok.map((m) => m.ttft_s).filter((v): v is number => v != null);
  const totals = ok.map((m) => m.total_s).filter((v): v is number => v != null);
  const tps = ok.map((m) => m.decode_tokens_per_s).filter((v): v is number => v != null);

  let faithfulnessChecked = 0;
  let faithfulnessPassed = 0;
  for (const m of ok) {
    const check = parseChecks(m.checks_json).find((c) => c.name === "faithfulness");
    if (!check) continue;
    faithfulnessChecked++;
    if (check.passed) faithfulnessPassed++;
  }

  return {
    endpoint_id: endpointId,
    case_id: caseId,
    n: scored.length,
    n_ok: ok.length,
    n_error: errors.length,
    ttft_p50: percentile(ttfts, 50),
    ttft_p95: percentile(ttfts, 95),
    ttft_p99: percentile(ttfts, 99),
    total_p50: percentile(totals, 50),
    total_p95: percentile(totals, 95),
    total_p99: percentile(totals, 99),
    decode_tps_mean: tps.length ? tps.reduce((a, b) => a + b, 0) / tps.length : null,
    faithfulness_checked: faithfulnessChecked,
    faithfulness_passed: faithfulnessPassed,
  };
}

export function summarizeAll(metrics: RequestMetrics[]): RunSummary[] {
  const groups = new Map<string, { endpointId: number; caseId: number; items: RequestMetrics[] }>();
  for (const m of metrics) {
    const key = `${m.endpoint_id}:${m.case_id}`;
    if (!groups.has(key)) groups.set(key, { endpointId: m.endpoint_id, caseId: m.case_id, items: [] });
    groups.get(key)!.items.push(m);
  }
  return [...groups.values()].map((g) => summarize(g.endpointId, g.caseId, g.items));
}

/** Tokens/sec during the decode phase only (excludes TTFT), so a slow
 * prompt-eval phase doesn't drag down the apparent generation speed.
 */
export function computeDecodeRate(
  completionTokens: number | null,
  ttftS: number | null,
  totalS: number | null
): number | null {
  if (!completionTokens || totalS == null || ttftS == null) return null;
  const decodeS = totalS - ttftS;
  if (decodeS <= 0) return null;
  return completionTokens / decodeS;
}
